File: rust/src/request_target.rs
//! Byte-level validation for raw registry request targets.

/// Maximum accepted raw request-target size.
pub const MAX_REQUEST_TARGET_BYTES: usize = 1024;
const MAX_JS_PACKAGE_BYTES: usize = 214;

/// A canonical origin-form path suitable for exact route-map lookup.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct CanonicalRequestTarget<'a> {
    path: &'a str,
}

impl<'a> CanonicalRequestTarget<'a> {
    /// Validates raw request-target bytes without decoding or normalizing them.
    #[must_use]
    pub fn parse(raw: &'a [u8]) -> Option<Self> {
        if raw.is_empty()
            || raw.len() > MAX_REQUEST_TARGET_BYTES
            || !raw.is_ascii()
            || raw.first() != Some(&b'/')
        {
            return None;
        }
        if raw == b"/" {
            return Some(Self { path: "/" });
        }
        if raw.last() == Some(&b'/') || raw.windows(2).any(|pair| pair == b"//") {
            return None;
        }
        if raw
            .split(|byte| *byte == b'/')
            .skip(1)
            .any(|segment| segment.is_empty() || segment == b"." || segment == b"..")
        {
            return None;
        }

        let valid = if raw.starts_with(b"/@") {
            valid_scoped_js_metadata_target(raw)
        } else {
            !raw.contains(&b'%') && raw[1..].iter().copied().all(valid_generic_path_byte)
        };
        if !valid {
            return None;
        }
        Some(Self {
            path: std::str::from_utf8(raw).ok()?,
        })
    }

    /// Returns the exact validated path bytes as UTF-8 text.
    #[must_use]
    pub const fn as_str(self) -> &'a str {
        self.path
    }
}

fn valid_generic_path_byte(byte: u8) -> bool {
    byte.is_ascii_alphanumeric() || matches!(byte, b'/' | b'.' | b'_' | b'~' | b'+' | b'@' | b'-')
}

fn valid_scoped_js_metadata_target(raw: &[u8]) -> bool {
    let encoded_name = &raw[2..];
    let Some(separator) = encoded_name.windows(3).position(|bytes| bytes == b"%2f") else {
        return false;
    };
    let scope = &encoded_name[..separator];
    let name = &encoded_name[separator + 3..];
    raw.len().checked_sub(3).is_some_and(|decoded_len| {
        decoded_len <= MAX_JS_PACKAGE_BYTES
            && !name.windows(3).any(|bytes| bytes == b"%2f")
            && valid_js_component(scope)
            && valid_js_component(name)
    })
}

fn valid_js_component(component: &[u8]) -> bool {
    component.first().is_some_and(u8::is_ascii_lowercase)
        && component[1..].iter().copied().all(|byte| {
            byte.is_ascii_lowercase()
                || byte.is_ascii_digit()
                || matches!(byte, b'.' | b'_' | b'~' | b'-')
        })
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;

    use serde::Deserialize;

    use super::*;

    const FIXTURE: &str = include_str!("../../fixtures/dynamic-registry-v1/http/raw-targets.json");

    #[derive(Debug, Deserialize)]
    #[serde(deny_unknown_fields)]
    #[serde(rename_all = "camelCase")]
    struct Fixture {
        cases: Vec<Case>,
        policy: Policy,
        schema: String,
    }

    #[derive(Debug, Deserialize)]
    #[serde(deny_unknown_fields)]
    #[serde(rename_all = "camelCase")]
    struct Policy {
        allowed_generic_segment_ascii: String,
        allowed_percent_escape: String,
        maximum_request_target_bytes: usize,
        scoped_java_script_component_pattern: String,
        scoped_java_script_maximum_package_bytes: usize,
        target_form: String,
    }

    #[derive(Debug, Deserialize)]
    #[serde(deny_unknown_fields)]
    #[serde(rename_all = "camelCase")]
    struct Case {
        expected: Expected,
        id: String,
        target_ascii: Option<String>,
        target_hex: Option<String>,
    }

    #[derive(Debug, Deserialize)]
    #[serde(deny_unknown_fields)]
    struct Expected {
        kind: ExpectedKind,
        path: Option<String>,
    }

    #[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq)]
    #[serde(rename_all = "kebab-case")]
    enum ExpectedKind {
        Canonical,
        Reject,
    }

    #[test]
    fn follows_shared_raw_target_vectors() {
        let fixture: Fixture = serde_json::from_str(FIXTURE).unwrap();
        assert_eq!(fixture.schema, "pkgre-raw-request-targets-v1");
        assert_eq!(
            fixture.policy.maximum_request_target_bytes,
            MAX_REQUEST_TARGET_BYTES
        );
        assert_eq!(
            fixture.policy.scoped_java_script_maximum_package_bytes,
            MAX_JS_PACKAGE_BYTES
        );
        assert_eq!(
            fixture.policy.allowed_generic_segment_ascii,
            "A-Z a-z 0-9 . _ ~ + @ -"
        );
        assert_eq!(
            fixture.policy.allowed_percent_escape,
            "one lowercase %2f in a canonical scoped JavaScript metadata path only"
        );
        assert_eq!(
            fixture.policy.scoped_java_script_component_pattern,
            "^[a-z][a-z0-9._~-]*$"
        );
        assert_eq!(
            fixture.policy.target_form,
            "origin-form path without query or fragment"
        );

        let mut ids = HashSet::new();
        for case in fixture.cases {
            assert!(ids.insert(case.id.clone()), "duplicate case {}", case.id);
            let raw = match (case.target_ascii, case.target_hex) {
                (Some(target), None) => target.into_bytes(),
                (None, Some(target)) => decode_hex(&target),
                _ => panic!("{} must define exactly one target encoding", case.id),
            };
            let actual = CanonicalRequestTarget::parse(&raw).map(CanonicalRequestTarget::as_str);
            match case.expected.kind {
                ExpectedKind::Canonical => {
                    assert_eq!(actual, case.expected.path.as_deref(), "case {}", case.id);
                }
                ExpectedKind::Reject => {
                    assert!(
                        case.expected.path.is_none(),
                        "reject case {} has a path",
                        case.id
                    );
                    assert_eq!(actual, None, "case {}", case.id);
                }
            }
        }
    }

    fn decode_hex(encoded: &str) -> Vec<u8> {
        assert_eq!(encoded.len() % 2, 0);
        encoded
            .as_bytes()
            .chunks_exact(2)
            .map(|pair| {
                let text = std::str::from_utf8(pair).unwrap();
                u8::from_str_radix(text, 16).unwrap()
            })
            .collect()
    }
}

File: js/src/request-target.js
import { Buffer } from "node:buffer";

export const MAX_REQUEST_TARGET_BYTES = 1024;
const MAX_JS_PACKAGE_BYTES = 214;

function validGenericPathByte(byte) {
  return (byte >= 0x30 && byte <= 0x39)
    || (byte >= 0x41 && byte <= 0x5a)
    || (byte >= 0x61 && byte <= 0x7a)
    || [0x2f, 0x2e, 0x5f, 0x7e, 0x2b, 0x40, 0x2d].includes(byte);
}

function validJsComponent(component) {
  if (!component.length || component[0] < 0x61 || component[0] > 0x7a) return false;
  return component.subarray(1).every((byte) => (byte >= 0x61 && byte <= 0x7a)
    || (byte >= 0x30 && byte <= 0x39)
    || [0x2e, 0x5f, 0x7e, 0x2d].includes(byte));
}

function encodedSlashOffset(bytes) {
  for (let index = 0; index + 2 < bytes.length; index += 1) {
    if (bytes[index] === 0x25 && bytes[index + 1] === 0x32 && bytes[index + 2] === 0x66) return index;
  }
  return -1;
}

function validScopedJsMetadataTarget(bytes) {
  const encodedName = bytes.subarray(2);
  const separator = encodedSlashOffset(encodedName);
  if (separator < 0 || encodedSlashOffset(encodedName.subarray(separator + 3)) >= 0) return false;
  return bytes.length - 3 <= MAX_JS_PACKAGE_BYTES
    && validJsComponent(encodedName.subarray(0, separator))
    && validJsComponent(encodedName.subarray(separator + 3));
}

/**
 * Validates raw request-target bytes without decoding or normalization.
 * @param {Uint8Array} raw
 * @returns {string | undefined}
 */
export function canonicalRequestTarget(raw) {
  if (!(raw instanceof Uint8Array)) throw new TypeError("raw request target must be a Uint8Array");
  const bytes = Buffer.from(raw.buffer, raw.byteOffset, raw.byteLength);
  if (!bytes.length || bytes.length > MAX_REQUEST_TARGET_BYTES || bytes[0] !== 0x2f || bytes.some((byte) => byte > 0x7f)) return undefined;
  if (bytes.length === 1) return "/";
  if (bytes.at(-1) === 0x2f) return undefined;

  let segmentStart = 1;
  for (let index = 1; index <= bytes.length; index += 1) {
    if (index !== bytes.length && bytes[index] !== 0x2f) continue;
    const segment = bytes.subarray(segmentStart, index);
    if (!segment.length || segment.equals(Buffer.from(".")) || segment.equals(Buffer.from(".."))) return undefined;
    segmentStart = index + 1;
  }

  const valid = bytes[1] === 0x40
    ? validScopedJsMetadataTarget(bytes)
    : !bytes.includes(0x25) && bytes.subarray(1).every(validGenericPathByte);
  return valid ? bytes.toString("ascii") : undefined;
}
